### PR TITLE
Add built-in topic to report the status of internal threads

### PR DESCRIPTION
**Pull request by jonesc-oci**

### CODE REVIEW

File: dds/DCPS/Discovery.cpp [hunk 1]
```diff
@@ -106,6 +106,43 @@ Discovery::create_bit_topics(DomainParticipantImpl* participant)
...
 
+  // Internal thread status topic
+  type_support =
+    Registered_Data_Types->lookup(participant, BUILT_IN_INTERNAL_THREAD_TOPIC_TYPE);
+
+  if (CORBA::is_nil(type_support)) {
+    OpenDDS::DCPS::InternalThreadBuiltinTopicDataTypeSupport_var ts =
+      new OpenDDS::DCPS::InternalThreadBuiltinTopicDataTypeSupportImpl;
+
+    DDS::ReturnCode_t ret = ts->register_type(participant,
```
Comment (jwillemsen) [inline dds/DCPS/Discovery.cpp:117]:
``` suggestion
    const DDS::ReturnCode_t ret = ts->register_type(participant,
```

File: dds/DCPS/DiscoveryBase.h [hunk 2]
```diff
@@ -93,17 +94,45 @@ namespace OpenDDS {
...
+      {
+        interval = TheServiceParticipant->get_thread_status_interval();
+        status = TheServiceParticipant->get_thread_statuses();
+        tid = ACE_OS::thr_self();
+        has_timeout = interval > TimeDuration(0);
+      }
 
       int svc()
       {
+        ACE_Time_Value expire = ACE_Time_Value(0);
```
Comment (jwillemsen) [inline dds/DCPS/DiscoveryBase.h:107]: Default constructor will initialise already to zero

File: dds/DCPS/DiscoveryBase.h [hunk 3]
```diff
@@ -93,17 +94,45 @@ namespace OpenDDS {
...
         {
           ACE_GUARD_RETURN(ACE_Thread_Mutex, g, mtx_, -1);
           reader_done_ = true;
           cnd_.signal();
           while (!writer_done_) {
-            cnd_.wait();
+            do {
+              cnd_.wait(&expire);
+
+              ACE_Time_Value now = MonotonicTimePoint::now().value();
```
Comment (jwillemsen) [inline dds/DCPS/DiscoveryBase.h:122]:
``` suggestion
             const ACE_Time_Value now = MonotonicTimePoint::now().value();
```

File: dds/DCPS/DiscoveryBase.h [hunk 4]
```diff
@@ -93,17 +94,45 @@ namespace OpenDDS {
...
+              if (has_timeout && now > expire) {
+                expire = now + interval.value();
+                if (status) {
+                  if (DCPS_debug_level > 4) {
+                    ACE_DEBUG((LM_DEBUG,
+                              "%T (%P|%t) DcpsUpcalls::svc. Updating thread status.\n"));
+                  }
+                  status->lock.acquire_write();
```
Comment (jwillemsen) [inline dds/DCPS/DiscoveryBase.h:130]: Use a guard instead of manual acquire/release?

File: dds/DCPS/DiscoveryBase.h [hunk 5]
```diff
@@ -128,6 +178,12 @@ namespace OpenDDS {
       bool reader_done_, writer_done_;
       ACE_Thread_Mutex mtx_;
       ACE_Condition_Thread_Mutex cnd_;
+
+      // thread reporting
+      TimeDuration interval;
+      ThreadStatus* status;
+      ACE_thread_t tid;
+      bool has_timeout;
```
Comment (jwillemsen) [inline dds/DCPS/DiscoveryBase.h:186]: Why have a seperate bool as member that indicates interval>0, use inline helper method? When you keep it, interval and has_timeout could be const when you initialise them directly

File: dds/DCPS/DiscoveryBase.h [hunk 6]
```diff
@@ -1822,6 +1878,16 @@ namespace OpenDDS {
...
     }
+
+    DCPS::InternalThreadBuiltinTopicDataDataReaderImpl* internal_thread_bit()
+    {
+      if (!bit_subscriber_.in())
+        return 0;
+
+      DDS::DataReader_var d =
+        bit_subscriber_->lookup_datareader(DCPS::BUILT_IN_INTERNAL_THREAD_TOPIC);
+      return dynamic_cast<InternalThreadBuiltinTopicDataDataReaderImpl*>(d.in());
```
Comment (jwillemsen) [inline dds/DCPS/DiscoveryBase.h:1889]: This looks risky, the _var will decrement the refcount when it goes out of scope, you are using the pointer after that.

File: dds/DCPS/RTPS/Spdp.cpp [hunk 7]
```diff
@@ -2147,6 +2149,23 @@ Spdp::SpdpTransport::open(const DCPS::ReactorTask_rch& reactor_task)
...
 #endif
 
+  // internal thread bit reporting
+  TimeDuration interval = TheServiceParticipant->get_thread_status_interval();
+
+  if (interval > TimeDuration(0)) {
+    if (DCPS::DCPS_debug_level > 4) {
+      ACE_DEBUG((LM_DEBUG,
+                ACE_TEXT("(%P|%t) Thread status monitoring is active. ")
+                ACE_TEXT("Status interval = %u.\n"), interval.value().sec()));
```
Comment (jwillemsen) [inline dds/DCPS/RTPS/Spdp.cpp:2159]: Why only log sec?

File: dds/DCPS/RTPS/Spdp.cpp [hunk 8]
```diff
@@ -3436,6 +3455,42 @@ void Spdp::SpdpTransport::send_local(const DCPS::MonotonicTimePoint& /*now*/)
...
+    ACE_DEBUG((LM_DEBUG,
+               "%T (%P|%t) Spdp::SpdpTransport::thread_status_task(): Updating internal thread status BIT.\n"));
+  }
+
+  const DCPS::RepoId guid = outer_->guid();
+  DCPS::InternalThreadBuiltinTopicDataDataReaderImpl* bit = outer_->internal_thread_bit();
+
+  if (TheServiceParticipant->get_thread_status_interval() > TimeDuration(0)) {
+    if (thread_status_ && bit) {
+      thread_status_->lock.acquire_read();
```
Comment (jwillemsen) [inline dds/DCPS/RTPS/Spdp.cpp:3470]: guard?

File: dds/DCPS/ReactorTask.h [hunk 9]
```diff
@@ -30,14 +30,22 @@ OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
 namespace OpenDDS {
 namespace DCPS {
 
+// thread status reporting forward
+struct ThreadStatus {
+  ACE_Recursive_Thread_Mutex lock;
```
Comment (jwillemsen) [inline dds/DCPS/ReactorTask.h:35]: Needs this to be recursive?

File: dds/DCPS/BuiltInTopicUtils.h [hunk 10]
```diff
@@ -39,6 +39,11 @@ OpenDDS_Dcps_Export extern const char* const BUILT_IN_PARTICIPANT_LOCATION_TOPIC
 OpenDDS_Dcps_Export extern const char* const BUILT_IN_CONNECTION_RECORD_TOPIC;
 OpenDDS_Dcps_Export extern const char* const BUILT_IN_CONNECTION_RECORD_TOPIC_TYPE;
 
+OpenDDS_Dcps_Export extern const char* const BUILT_IN_INTERNAL_THREAD_TOPIC;
+OpenDDS_Dcps_Export extern const char* const BUILT_IN_INTERNAL_THREAD_TOPIC_TYPE;
+
+enum { NUMBER_OF_BUILT_IN_TOPICS = 7 };
```
Comment (iguessthislldo) [inline dds/DCPS/BuiltInTopicUtils.h:45]:
```suggestion
const size_t NUMBER_OF_BUILT_IN_TOPICS = 7;
```

File: dds/DCPS/Discovery.cpp [hunk 11]
```diff
@@ -106,6 +106,43 @@ Discovery::create_bit_topics(DomainParticipantImpl* participant)
...
+  }
+
+  DDS::Topic_var bit_internal_thread_topic =
+    participant->create_topic(BUILT_IN_INTERNAL_THREAD_TOPIC,
+      BUILT_IN_INTERNAL_THREAD_TOPIC_TYPE,
+      TOPIC_QOS_DEFAULT,
+      DDS::TopicListener::_nil(),
+      DEFAULT_STATUS_MASK);
+
+  if (CORBA::is_nil(bit_internal_thread_topic)) {
```
Comment (iguessthislldo) [inline dds/DCPS/Discovery.cpp:137]:
```suggestion
  if (!bit_internal_thread_topic) {
```

File: tests/DCPS/InternalThreadStatus/InternalThreadStatusListenerImpl.h [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+{
+public:
+  //Constructor
+  explicit InternalThreadStatusListenerImpl(OPENDDS_STRING id);
+
+  //Destructor
+  virtual ~InternalThreadStatusListenerImpl();
+
+  virtual void on_requested_deadline_missed(
+                                            DDS::DataReader_ptr reader,
```
Comment (iguessthislldo) [inline tests/DCPS/InternalThreadStatus/InternalThreadStatusListenerImpl.h:33]:
```suggestion
  virtual void on_requested_deadline_missed(DDS::DataReader_ptr reader,
```

File: dds/DCPS/ReactorTask.h [hunk 13]
```diff
@@ -30,14 +30,22 @@ OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
...
+  OPENDDS_MAP(unsigned long, ACE_Time_Value) map;
```
Comment (iguessthislldo) [inline dds/DCPS/ReactorTask.h:36]: Please avoid dealing with `ACE_Time_Value`s unless you're dealing directly with ACE.

**Review by iguessthislldo — CHANGES_REQUESTED** [review]

File: dds/DCPS/transport/framework/QueueTaskBase_T.h [hunk 14]
```diff
@@ -240,6 +246,9 @@ class QueueTaskBase : public ACE_Task_Base {
 
   /// The id of the thread created by this task.
   ACE_thread_t thr_id_;
+
+  /// name for trhead monitoring BIT
```
Comment (iguessthislldo) [inline dds/DCPS/transport/framework/QueueTaskBase_T.h:250]:
```suggestion
  /// name for thread monitoring BIT
```

File: dds/DCPS/RTPS/Spdp.cpp [hunk 15]
```diff
@@ -3436,6 +3461,44 @@ void Spdp::SpdpTransport::send_local(const DCPS::MonotonicTimePoint& /*now*/)
   outer_->remove_expired_participants();
 }
 
+void Spdp::SpdpTransport::thread_status_task(const DCPS::MonotonicTimePoint& /*now*/)
```
Comment (jrw972) [inline dds/DCPS/RTPS/Spdp.cpp:3464]: What challenges are there to moving the reporting logic into the Service Participant or at least the DomainParticipantImpl?  That way, it generalizes to all types of discovery.  Alternatively, we could forego the thread status map and have each thing that would be inserting into the map just write its instance when necessary.

I'm not interested is doing this right now but I would like to know if there are certain challenges to doing so.

**Review by jrw972 — CHANGES_REQUESTED** [review]

File: dds/DCPS/DiscoveryBase.h [hunk 16]
```diff
@@ -86,24 +87,70 @@ namespace OpenDDS {
...
+                expire = now.value() + interval.value();
+                if (status) {
+                  if (DCPS_debug_level > 4) {
+                    ACE_DEBUG((LM_DEBUG,
+                              "%T (%P|%t) DcpsUpcalls::svc. Updating thread status.\n"));
+                  }
+                  {
+                    ACE_WRITE_GUARD_RETURN(ACE_Thread_Mutex, g, status->lock, -1);
+                    status->map[key] = now;
+                  }
```
Comment (mitza-oci) [inline dds/DCPS/DiscoveryBase.h:150]:
```suggestion
                  ACE_WRITE_GUARD_RETURN(ACE_Thread_Mutex, g, status->lock, -1);
                  status->map[key] = now;
```

File: dds/DCPS/DiscoveryBase.h [hunk 17]
```diff
@@ -86,24 +87,70 @@ namespace OpenDDS {
...
+        }
+
         drr_->add_association(reader_, wa_, active_);
         {
           ACE_GUARD_RETURN(ACE_Thread_Mutex, g, mtx_, -1);
           reader_done_ = true;
           cnd_.signal();
           while (!writer_done_) {
-            cnd_.wait();
+            do {
```
Comment (mitza-oci) [inline dds/DCPS/DiscoveryBase.h:136]: Can these be combined?

File: dds/DCPS/DiscoveryBase.h [hunk 18]
```diff
@@ -86,24 +87,70 @@ namespace OpenDDS {
...
+              cnd_.wait(&expire);
+
+              MonotonicTimePoint now = MonotonicTimePoint::now();
```
Comment (mitza-oci) [inline dds/DCPS/DiscoveryBase.h:139]:
```suggestion
              const MonotonicTimePoint now = MonotonicTimePoint::now();
```

File: dds/DCPS/RTPS/Spdp.cpp [hunk 19]
```diff
@@ -3436,6 +3461,43 @@ void Spdp::SpdpTransport::send_local(const DCPS::MonotonicTimePoint& /*now*/)
...
+    ACE_DEBUG((LM_DEBUG,
+               "%T (%P|%t) Spdp::SpdpTransport::thread_status_task(): Updating internal thread status BIT.\n"));
+  }
+
+  const DCPS::RepoId guid = outer_->guid();
+  DCPS::InternalThreadBuiltinTopicDataDataReaderImpl* bit = outer_->internal_thread_bit();
+
+  if (TheServiceParticipant->get_thread_status_interval() > TimeDuration(0)) {
+    if (thread_status_ && bit) {
+      {
```
Comment (mitza-oci) [inline dds/DCPS/RTPS/Spdp.cpp:3477]:
```suggestion
```

File: dds/DCPS/transport/framework/TransportImpl.cpp [hunk 20]
```diff
@@ -106,14 +106,15 @@ TransportImpl::add_pending_connection(const TransportClient_rch& client, DataLin
 }
 
 void
-TransportImpl::create_reactor_task(bool useAsyncSend)
+TransportImpl::create_reactor_task(bool useAsyncSend, const OPENDDS_STRING name)
```
Comment (mitza-oci) [inline dds/DCPS/transport/framework/TransportImpl.cpp:109]:
```suggestion
TransportImpl::create_reactor_task(bool useAsyncSend, const OPENDDS_STRING& name)
```

File: dds/DCPS/DiscoveryBase.h [hunk 21]
```diff
@@ -86,24 +87,70 @@ namespace OpenDDS {
     }
 
     struct DcpsUpcalls : ACE_Task_Base {
+      inline bool has_timeout() {
```
Comment (mitza-oci) [inline dds/DCPS/DiscoveryBase.h:90]:
```suggestion
      bool has_timeout() const {
```
already inline (defined within class definition)

**Review by mitza-oci — CHANGES_REQUESTED** [review]

**Review by mitza-oci — CHANGES_REQUESTED** [review]

**Review by mitza-oci — APPROVED** [review]

**Pull request merged**